community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/AcceptingExecutionResult.scala

```diff
@@ -0,0 +1,270 @@
+ */
+package org.neo4j.cypher.internal.compiler.v3_0.executionplan
+
+import java.io.{PrintWriter, StringWriter}
+import java.util
+
+import org.neo4j.cypher.internal.compiler.v3_0.commands.values.KeyToken
+import org.neo4j.cypher.internal.compiler.v3_0.helpers.IsCollection
+import org.neo4j.cypher.internal.compiler.v3_0.planDescription.InternalPlanDescription
+import org.neo4j.cypher.internal.compiler.v3_0.planDescription.InternalPlanDescription.Arguments.{Planner, Runtime}
+import org.neo4j.cypher.internal.compiler.v3_0.spi.{InternalResultRow, InternalResultVisitor, QueryContext}
+import org.neo4j.cypher.internal.compiler.v3_0.{TaskCloser, _}
+import org.neo4j.cypher.internal.frontend.v3_0.EntityNotFoundException
+import org.neo4j.cypher.internal.frontend.v3_0.helpers.Eagerly
+import org.neo4j.cypher.internal.frontend.v3_0.notification.InternalNotification
+import org.neo4j.graphdb.{ResourceIterator, _}
+
+import scala.collection.{Map, mutable}
+
+/*
+* An AcceptingExecutionResult delegates all methods to the accept method which is the responsibility
+* of extending classes to provide.
+*/
+abstract class AcceptingExecutionResult(context: QueryContext,
+                                        taskCloser: Option[TaskCloser] = None)
+  extends InternalExecutionResult with SuccessfulCloseable {
+  self =>
+
+  import scala.collection.JavaConverters._
+
+  private lazy val innerIterator: util.Iterator[util.Map[String, Any]] = {
+    val list = new util.ArrayList[util.Map[String, Any]]()
+    if (isOpen) doInAccept(populateResults(list))
+    list.iterator()
+  }
+
+  override def columns: List[String] = javaColumns.asScala.toList
+
+  override final def javaColumnAs[T](column: String): ResourceIterator[T] = new WrappingResourceIterator[T] {
+    def hasNext = self.hasNext
+
+    def next() = extractJavaColumn(column, innerIterator.next()).asInstanceOf[T]
+  }
+
+  override def columnAs[T](column: String): Iterator[T] = map { case m => extractColumn(column, m).asInstanceOf[T] }
+
+  override def javaIterator: ResourceIterator[util.Map[String, Any]] = new
+      WrappingResourceIterator[util.Map[String, Any]] {
+    def hasNext = self.hasNext
+
+    def next() = innerIterator.next()
+  }
+
+  override def dumpToString(): String = {
+    val stringWriter = new StringWriter()
+    val writer = new PrintWriter(stringWriter)
+    dumpToString(writer)
+    writer.close()
+    stringWriter.getBuffer.toString
+  }
+
+  override def dumpToString(writer: PrintWriter) = {
+    val builder = Seq.newBuilder[Map[String, String]]
+    doInAccept(populateDumpToStringResults(builder))
+    formatOutput(writer, columns, builder.result(), queryStatistics())
+  }
+
+  /*
+   * NOTE: This should ony be used for testing, it creates an InternalExecutionResult
+   * where you can call both toList and dumpToString
+   */
+  def toEagerIterableResult(planner: PlannerName, runtime: RuntimeName): InternalExecutionResult = {
```

Comment:
it needs direct access to `doInAccept`. This is only used to make it possible to create `RewindableExecutionResult` for tests. We should refactor how that works and remove this method instead.